### PR TITLE
wgengine/winnet: invoke some COM methods directly instead of through …

### PR DESCRIPTION
**Pull request by dblohm7**
…IDispatch.

Intermittently in the wild we are seeing failures when calling `INetworkConnection::GetNetwork`. It is unclear what the root cause is, but what is clear is that the error is happening inside the object's `IDispatch` invoker (as opposed to the method implementation itself).

This patch replaces our wrapper for `INetworkConnection::GetNetwork` with an alternate implementation that directly invokes the method, instead of using `IDispatch`. I also replaced the implementations of `INetwork::SetCategory` and `INetwork::GetCategory` while I was there.

This patch is speculative and tightly-scoped so that we could possibly add it to a dot-release if necessary.

Updates https://github.com/tailscale/tailscale/issues/6037

Signed-off-by: Aaron Klotz <aaron@tailscale.com>

### CODE REVIEW

**Comment by bradfitz** [comment]
Updates #4134 too?

File: wgengine/winnet/winnet.go [hunk 1]
```diff
@@ -45,6 +46,23 @@ type INetwork struct {
...
+	GetNetworkId               uintptr
+	GetDomainType              uintptr
+	GetNetworkConnections      uintptr
+	GetTimeCreatedAndConnected uintptr
+	Get_IsConnectedToInternet  uintptr
+	Get_IsConnected            uintptr
+	GetConnectivity            uintptr
+	GetCategory                uintptr
+	SetCategory                uintptr
+}
```
Comment (andrew-d) [inline wgengine/winnet/winnet.go:64]: This appears to match the vtable that Microsoft generates for their Rust API bindings (my go-to for finding the structure for things without needing to dig through C++):
https://github.com/microsoft/windows-rs/blob/962d0a9f19d049a37f704b8ea3c7ab7efd1d95cb/crates/libs/windows/src/Windows/Win32/Networking/NetworkListManager/mod.rs#L248-L269

**Review by andrew-d — COMMENTED** [review]
This is the first time that I'm looking at our COM code; super neat stuff 👍

**Comment by dblohm7** [comment]
> Updates #4134 too?

Yeah. Fixed.

**Comment by dblohm7** [comment]
> This is the first time that I'm looking at our COM code; super neat stuff 👍

TBH I despise the library we're using for this and am working on something much better.

**Comment by raggi** [comment]
I made a basic debugging test, and I'm still seeing the same behavior with the patch:

```
func TestBasic(t *testing.T) {

	runtime.LockOSThread()
	defer runtime.UnlockOSThread()

	var c ole.Connection
	if err := c.Initialize(); err != nil {
		t.Fatal(err)
	}
	defer c.Uninitialize()

	m, err := NewNetworkListManager(&c)
	if err != nil {
		t.Fatal(err)
	}
	defer m.Release()

	cl, err := m.GetNetworkConnections()
	if err != nil {
		t.Fatal(err)
	}
	defer cl.Release()

	t.Logf("got %d connections", len(cl))
	for _, c := range cl {
		conn, err := c.GetNetwork()
		if err != nil {
			t.Errorf("GetNetwork: %v", err)
			continue
		}
		name, err := conn.GetName()
		if err != nil {
			t.Errorf("GetName: %v", err)
			continue
		} else {
			t.Logf("name: %s", name)
		}
		id, err := c.GetAdapterId()
		if err != nil {
			t.Fatal(err)
		}
		t.Logf("id: %s", id)
	}

}
```

```
=== RUN   TestBasic
    wininet_test.go:36: got 2 connections
    wininet_test.go:45: GetName: Element not found.
    wininet_test.go:45: GetName: Element not found.
--- FAIL: TestBasic (0.02s)
FAIL
FAIL    tailscale.com/wgengine/winnet   0.278s
FAIL
```

**Comment by dblohm7** [comment]
As discussed with James, the same code does not repro on my machine. We're not calling `GetName` in this code path, so I'd like to land the vtable stuff for now so that we can give support an unstable build.

**Review by raggi — APPROVED** [review]
The code here is fine, and works in principle - following the pattern of other usage of the lib (e.g. the bluetooth library that uses it). It still doesn't seem to work on my Windows 11, but I think it's a step in the right direction anyway.

**Pull request merged**